iree/compiler/Dialect/Flow/Transforms/test/dispatch_id_canonicalization.mlir

```diff
@@ -0,0 +1,76 @@
+// RUN: iree-opt -split-input-file -test-iree-flow-dispatch-id-canonicalizations %s | IreeFileCheck %s
+
+#map0 = affine_map<(d0) -> (d0 * 2)>
+
+// CHECK-LABEL: func @single_iter_for
+//   CHECK-NOT:   scf.for
+//       CHECK:   flow.return
+func @single_iter_for(%arg0: tensor<8x8xf32>) -> tensor<8x8xf32> {
+  %c4 = constant 4 : index
+  %3 = flow.dispatch.workgroups[%c4, %c4, %c4] (%arg0) : (tensor<8x8xf32>)
+    -> tensor<8x8xf32> = (%arg1 : !flow.dispatch.input<8x8xf32>, %arg2 : !flow.dispatch.output<8x8xf32>) {
+    %c1 = constant 1 : index
+    %c2 = constant 2 : index
+    %c8 = constant 8 : index
+    %workgroup_id_2 = flow.dispatch.workgroup.id[2] : index
+    %workgroup_count_2 = flow.dispatch.workgroup.count[2] : index
+    %4 = affine.apply #map0(%workgroup_id_2)
+    %5 = affine.apply #map0(%workgroup_count_2)
+    scf.for %arg4 = %4 to %c8 step %5 {
```

Comment:
We should test this with dynamic shapes as well - one thing I'd want to make sure of - given where this runs in the pipeline (inline during the construction) - is that if a dynamic shape dimension is passed in as an argument and not constant at this point this is still useful. If it's not, it may be better to run this after outlining on the flow.executable where the dynamic dimensions have been materialized and passes like RematerializeDispatchConstants (which may inline some of those dynamic dimensions into the region after other passes run) have a chance to run. Today with where this is running if a dynamic dimension gets turned into a constant this may behave differently -- or may not - would be useful to test :)

Reply:
Thanks for pointing that out. This transformation won't help much if the dispatch sizes are none-constant. The reason I ran it at this stage is because otherwise the `flow.dispatch.input.load` ops don't seem to get transformed to static shape ops but I think we should be able to canonicalize later one.
I'll make those transformations run separately later on then.

Reply:
I moved the pass after `RematerializeDispatchConstants` but I'm not sure this is the right spot. After outlining I cannot easily access the dispatch count anymore right? I'm probably missing something.

Reply:
> This transformation won't help much if the dispatch sizes are none-constant.

ah ok thanks for explaining! That would be useful to note in the pass description ("...when the dispatch count is constant"). Renaming the pass would also help indicate when it's used more clearly: `PropagateConstantDispatchCounts` or something like it (following the pattern of "verb adjective noun" to make it clear what's being done to what and when, and to someone looking at the pass list make more sense as to when it should run). The more general 'canonicalization' name makes me think this is doing more than that and more generally useful.

I'm interested to know how load-bearing this is: is this strictly IR cleanup to make things easier to read (which is totally fine!) or does this have a significant impact? Anything making assumptions about the dispatch count is at the flow level generally an anti-pattern when dynamic shapes are involved. It may even be invalid for this kind of optimization even in static cases: workgroup sizes are not yet decided as until we go into the HAL we intentionally are target backend agnostic, meaning that a workgroup count of x=50 here in flow may end up being 2 in hal if the workgroup size for a particular backend is chosen to be 25, etc.

Reply:
I agree, I missed the part of where it is added to the pipeline. I would actually not make any decisions here at Flow level w.r.t to number of workgroups, etc. When lowering from Flow to HAL you run the codegen passes on the computation. It should be done there....

Reply:
> is this strictly IR cleanup to make things easier to read (which is totally fine!) or does this have a significant impact?

It is an optimization to remove useless control flow when the dispatch size is static and this also allows vectorization to work as it currently require static shape.

>I agree, I missed the part of where it is added to the pipeline. I would actually not make any decisions here at Flow level w.r.t to number of workgroups, etc. When lowering from Flow to HAL you run the codegen passes on the computation. It should be done there....

Mahesh, so you are saying this is not using the right dialect? What Ops should it be based on?

Reply:
If I am not mistaken there are corresponding operations in HAL dialect as well. Ideally I would trigger this after we have `LaunchConfig`. Because that is determining the static tile size to use, and the workgroup size. Then you also have the number of workgroups. That is where I was thinking this could slot in. Am I missing something?

Reply:
I see so I should be using `hal.interface.workgroup.id` instead? Can I still access the number of workgroup dispatched in this case? It seemed like it got moved into the function caller?

Reply:
Maybe you are right about that. I have to still get upto speed on that (backlog flushing some failures on patches that are ready to land). Are you planning to do vectorization at flow level. That would be problematic in terms of making things work with everything else that already exists.

Reply:
We should def. aim at making this work with non-constant quantities in the future.
e.g. `for i = 0; i < ub; i += workgroup_size` where we also know that `ub == cst * workgroup_size` or even just `ub % workgroup_size == 0`.
No need to make logic more complex now but let's add a TODO please.